[scalardb-cluster] Support Encryption in ScalarDB Cluster chart

kota2and3kan opened the pull request:
## Description

This PR adds new features to support the TDE feature of ScalarDB Cluster.

At this time, there are three options to use the TDE feature of ScalarDB Cluster.

1. Using Vault to manage the DEK (data encryption key) with non-TLS connections.

   ```
   [Vault] <---(non-TLS / encrypt and decrypt API)--- [ScalarDB Cluster] ---(Put or get encrypted data)---> [Backend DB]
   ```

2. Using Vault to manage the DEK (data encryption key) with TLS connections.

   ```
   [Vault] <---(TLS / encrypt and decrypt API)--- [ScalarDB Cluster] ---(Put or get encrypted data)---> [Backend DB]
   ```

3. Using Kubernetes Secret to manage the DEK (data encryption key).

   ```
   [Kubernetes Secret] <---(create, store, and get the data encryption key)--- [ScalarDB Cluster] ---(Put or get encrypted data)---> [Backend DB]
   ```

This PR adds the new features to support the above three options as follows:

1. Using Vault to manage the DEK (data encryption key) with non-TLS connections.
   - In this option, basically, this chart doesn't need additional configurations.
     - ScalarDB Cluster connects to Vault based on the configurations of ScalarDB Cluster itself.
2. Using Vault to manage the DEK (data encryption key) with TLS connections.
   - In this option, this chart should accept secret resource to mount the CA root certificate to access Vault via TLS. 
     - You can enable TLS-related features for Vault by setting `scalardbCluster.tde.vault.tls.enabled=true`.
     - You can specify the secret resource name to `scalardbCluster.tde.vault.tls.caRootCertSecret`.
3. Using Kubernetes Secret to manage the DEK (data encryption key).
   - In this option, this chart must allow ScalarDB Cluster pod to run the several API to create, get, and delete secret resources.
     - You can set necessary permissions to the role resource by setting `scalardbCluster.tde.type=self`.

Please take a look!

## Related issues and/or PRs

- You can see the documents related to this PR in the following PR.
  - https://github.com/scalar-labs/docs-internal-orchestration/pull/69
- The TDE features are implemented on the ScalarDB Cluster side.
  - https://github.com/scalar-labs/scalardb-cluster/pull/701
  - https://github.com/scalar-labs/scalardb-cluster/pull/734

## Changes made

- Add TDE related configurations and resources.

## Checklist

- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation to reflect the changes.
- [x] Any remaining open issues linked to this PR are documented and up-to-date (Jira, GitHub, etc.).
- [x] Tests (unit, integration, etc.) have been added for the changes.
- [x] My changes generate no new warnings.
- [x] Any dependent changes in other PRs have been merged and published.

## Additional notes (optional)

N/A

## Release notes

Add TDE support of ScalarDB Cluster. You can deploy ScalarDB Cluster with TDE feature by using new helm chart.

kota2and3kan commented on charts/scalardb-cluster/templates/scalardb-cluster/deployment.yaml:
```diff
@@ -52,6 +52,10 @@ spec:
             value: {{ include "scalardb-cluster.fullname" . }}-headless
           - name: SCALAR_DB_CLUSTER_LOG_LEVEL
             value: "{{ .Values.scalardbCluster.logLevel }}"
+          {{- if and (.Values.scalardbCluster.tde.enabled) (eq .Values.scalardbCluster.tde.type "self") }}
+          - name: SCALAR_DB_CLUSTER_ENCRYPTION_SELF_KUBERNETES_SECRET_NAMESPACE_NAME
+            value: {{ .Release.Namespace }}
+          {{- end }}
```
If we use the `self` type TDE, we should set the namespace name automatically via an environment variable. This is because ScalarDB Cluster uses this information to run the Kubernetes API to create, get, and delete the secret resources to manage the data encryption key.

kota2and3kan commented on charts/scalardb-cluster/templates/scalardb-cluster/deployment.yaml:
```diff
@@ -97,6 +101,10 @@ spec:
             - name: scalardb-cluster-tls-volume
               mountPath: /tls/scalardb-cluster/certs
             {{- end }}
+            {{- if and (.Values.scalardbCluster.tde.enabled) (eq .Values.scalardbCluster.tde.type "vault") (.Values.scalardbCluster.tde.vault.tls.enabled) }}
+            - name: scalardb-cluster-tde-vault-tls-volume
+              mountPath: /tde/vault/tls/certs
+            {{- end }}
```
If we use the `vault` type TDE and use TLS between ScalarDB Cluster and Vault, we can set a CA root certificate to access to Vault from ScalarDB Cluster by using a secret resource. This chart mounts the CA root certificate on pods from the secret resource.

kota2and3kan commented on charts/scalardb-cluster/templates/scalardb-cluster/role.yaml:
```diff
@@ -7,3 +7,8 @@ rules:
   - apiGroups: [""]
     resources: ["endpoints"]
     verbs: ["get", "watch", "list"]
+  {{- if and (.Values.scalardbCluster.tde.enabled) (eq .Values.scalardbCluster.tde.type "self") }}
+  - apiGroups: [""]
+    resources: ["secrets"]
+    verbs: ["get", "create", "delete"]
+  {{- end }}
```
If we use the `self` type TDE, we must allow ScalarDB Cluster's pods to create, get, and delete the Kubernetes Secrets resources. This is because ScalarDB Cluster manages the data encryption keys by using the Kubernetes Secrets.

komamitsu reviewed (approved):
LGTM! 👍

Torch3333 commented on charts/scalardb-cluster/values.yaml:
```diff
@@ -317,3 +317,17 @@ scalardbCluster:
       issuerRef: {}
+
+  tde:
+    # -- Enable TDE. You must enable TDE when you use the TDE feature of  ScalarDB Cluster.
```
```suggestion
    # -- Enable TDE. You must enable TDE when you use the TDE feature of ScalarDB Cluster.
```

kota2and3kan replied:
Good catch! Thank you!

Torch3333 reviewed (approved):
LGTM, thank you!

feeblefakie commented:
@kota2and3kan Sorry, we probably don't use the proper noun `ScalarDB TDE` for this feature since we will name ScalarDB features simply with common names.
Please see the comment and around it for the background.
Let us discuss it and keep you updated later.

kota2and3kan commented:
@feeblefakie  
I see. I understood.
Thank you for sharing!

kota2and3kan commented:
@feeblefakie @brfrn169 
As we discussed in the meeting, I updated the term `TDE` to `Encryption`.
Please take a look when you have time! 🙇‍♀️

josh-wong commented on charts/scalardb-cluster/README.md:
```diff
@@ -25,6 +25,12 @@ Current chart version is `2.0.0-SNAPSHOT`
 | fullnameOverride | string | `""` | String to fully override scalardb-cluster.fullname template |
 | nameOverride | string | `""` | String to partially override scalardb-cluster.fullname template (will maintain the release name) |
 | scalardbCluster.affinity | object | `{}` | The affinity/anti-affinity feature, greatly expands the types of constraints you can express. |
+| scalardbCluster.encryption.enabled | bool | `false` | Enable encryption feature. You must enable encryption when you use the encryption feature of ScalarDB Cluster. |
```
```suggestion
| scalardbCluster.encryption.enabled | bool | `false` | Enable the encryption feature. To use encryption in ScalarDB Cluster, you must enable the encryption feature. |
```

josh-wong commented on charts/scalardb-cluster/README.md:
```diff
@@ -25,6 +25,12 @@ Current chart version is `2.0.0-SNAPSHOT`
 | nameOverride | string | `""` | String to partially override scalardb-cluster.fullname template (will maintain the release name) |
 | scalardbCluster.affinity | object | `{}` | The affinity/anti-affinity feature, greatly expands the types of constraints you can express. |
+| scalardbCluster.encryption.enabled | bool | `false` | Enable encryption feature. You must enable encryption when you use the encryption feature of ScalarDB Cluster. |
+| scalardbCluster.encryption.type | string | `""` | Type of encryption. You must set the same value as "scalar.db.cluster.encryption.type" of ScalarDB Cluster. |
```
```suggestion
| scalardbCluster.encryption.type | string | `""` | Type of encryption. You must set this value to the same value as "scalar.db.cluster.encryption.type" for ScalarDB Cluster. |
```

feeblefakie commented on charts/scalardb-cluster/README.md:
```diff
@@ -25,6 +25,12 @@ Current chart version is `2.0.0-SNAPSHOT`
+| scalardbCluster.encryption.enabled | bool | `false` | Enable encryption feature. You must enable encryption when you use the encryption feature of ScalarDB Cluster. |
+| scalardbCluster.encryption.type | string | `""` | Type of encryption. You must set the same value as "scalar.db.cluster.encryption.type" of ScalarDB Cluster. |
+| scalardbCluster.encryption.vault | object | `{"tls":{"caRootCertSecret":"","enabled":false}}` | Vault specific configurations. |
+| scalardbCluster.encryption.vault.tls | object | `{"caRootCertSecret":"","enabled":false}` | TLS configurations to access from ScalarDB Cluster to Vault by using TLS. |
```
It's more like a question to @brfrn169.
I don't see TLS-related configurations in the current doc, but are they going to be added?
https://github.com/scalar-labs/docs-internal-scalardb/pull/452

kota2and3kan replied:
Ah, I see. I overlooked it.
I will create a PR to add those TLS configurations!

kota2and3kan replied:
I created the PRs!
- https://github.com/scalar-labs/docs-internal-scalardb/pull/488
- https://github.com/scalar-labs/docs-internal-scalardb/pull/489

brfrn169 replied:
@kota2and3kan Thanks!

feeblefakie reviewed (commented):
Overall, looking good.
Left a suggestion and a question.

josh-wong commented on charts/scalardb-cluster/README.md:
```diff
@@ -25,6 +25,12 @@ Current chart version is `2.0.0-SNAPSHOT`
 | scalardbCluster.affinity | object | `{}` | The affinity/anti-affinity feature, greatly expands the types of constraints you can express. |
+| scalardbCluster.encryption.enabled | bool | `false` | Enable encryption feature. You must enable encryption when you use the encryption feature of ScalarDB Cluster. |
+| scalardbCluster.encryption.type | string | `""` | Type of encryption. You must set the same value as "scalar.db.cluster.encryption.type" of ScalarDB Cluster. |
+| scalardbCluster.encryption.vault | object | `{"tls":{"caRootCertSecret":"","enabled":false}}` | Vault specific configurations. |
```
```suggestion
| scalardbCluster.encryption.vault | object | `{"tls":{"caRootCertSecret":"","enabled":false}}` | Vault-specific configurations. |
```

josh-wong commented on charts/scalardb-cluster/README.md:
```diff
@@ -25,6 +25,12 @@ Current chart version is `2.0.0-SNAPSHOT`
+| scalardbCluster.encryption.type | string | `""` | Type of encryption. You must set the same value as "scalar.db.cluster.encryption.type" of ScalarDB Cluster. |
+| scalardbCluster.encryption.vault | object | `{"tls":{"caRootCertSecret":"","enabled":false}}` | Vault specific configurations. |
+| scalardbCluster.encryption.vault.tls | object | `{"caRootCertSecret":"","enabled":false}` | TLS configurations to access from ScalarDB Cluster to Vault by using TLS. |
+| scalardbCluster.encryption.vault.tls.caRootCertSecret | string | `""` | Name of the Secret containing the custom CA root certificate for TLS communication between ScalarDB Cluster and Vault. The certificate file will be mounted under the /encryption/vault/tls/certs/ directory in the ScalarDB Cluster pod. |
```
```suggestion
| scalardbCluster.encryption.vault.tls.caRootCertSecret | string | `""` | Name of the Secret containing the custom CA root certificate for TLS communication between ScalarDB Cluster and Vault. The certificate file will be mounted under the `/encryption/vault/tls/certs/` directory in the ScalarDB Cluster pod. |
```

josh-wong commented on charts/scalardb-cluster/values.yaml:
```diff
@@ -317,3 +317,17 @@ scalardbCluster:
       issuerRef: {}
+
+  encryption:
+    # -- Enable encryption feature. You must enable encryption when you use the encryption feature of ScalarDB Cluster.
+    enabled: false
+    # -- Type of encryption. You must set the same value as "scalar.db.cluster.encryption.type" of ScalarDB Cluster.
+    type: ""
+    # -- Vault specific configurations.
+    vault:
+      # -- TLS configurations to access from ScalarDB Cluster to Vault by using TLS.
+      tls:
+        # -- Enable TLS between ScalarDB Cluster and Vault. Note that you must enable the TLS feature on the Vault side.
+        enabled: false
+        # -- Name of the Secret containing the custom CA root certificate for TLS communication between ScalarDB Cluster and Vault. The certificate file will be mounted under the /encryption/vault/tls/certs/ directory in the ScalarDB Cluster pod.
```
Please update the comments in this section based on the changes to the tables with the same wording in `charts/scalardb-cluster/README.md`. I didn't make suggestions for each one here, just in case we need to change the wording that I suggested.

josh-wong reviewed (changes requested):
I've added some comments and suggestions. PTAL!

kota2and3kan commented:
Thank you for your review!

@feeblefakie  
I applied your suggestion and left the answer to your question (created other PRs).

@josh-wong  
I applied your suggestions. Actually, the `README.md` is created automatically based on the `values.yaml` file. So, I updated the `values.yaml` file, and ran a script to generate the `README.md`, instead of applying your suggestions to the `README.md` directly.

Please take a look again when you have time!

brfrn169 reviewed (approved):
LGTM! Thank you!

kota2and3kan commented:
@feeblefakie @brfrn169 
As we discussed in the meeting, I added a note about HPC Vault Dedicated in the `README.md`.

@josh-wong  
I updated the `README.md` in 7f9634d95cce9a7794ad013299766ad71baefa98. So, could you please take a look at the new comments? 🙇‍♂️

josh-wong commented on charts/scalardb-cluster/README.md:
```diff
@@ -25,6 +25,12 @@ Current chart version is `2.0.0-SNAPSHOT`
+| scalardbCluster.encryption.enabled | bool | `false` | Enable encryption at rest. You must set this to `true` if you're using the encryption feature in ScalarDB Cluster. |
+| scalardbCluster.encryption.type | string | `""` | Type of encryption. You must set this value to the same value as "scalar.db.cluster.encryption.type" for ScalarDB Cluster. |
+| scalardbCluster.encryption.vault | object | `{"tls":{"caRootCertSecret":"","enabled":false}}` | Vault-specific configurations. |
+| scalardbCluster.encryption.vault.tls | object | `{"caRootCertSecret":"","enabled":false}` | TLS configurations to access from ScalarDB Cluster to Vault by using TLS. When you use HCP Vault Dedicated, you don't need to set these TLS configurations even though HCP Vault Dedicated uses HTTPS. You need to set these TLS configurations if and only if you use other HashiCorp Vault deployments than HCP ...
```
```suggestion
| scalardbCluster.encryption.vault.tls | object | `{"caRootCertSecret":"","enabled":false}` | TLS configurations to provide access from ScalarDB Cluster to Vault by using TLS. If you're using HashiCorp Cloud Platform (HCP) Vault Dedicated, you don't need to set these TLS configurations even though HCP Vault Dedicated uses HTTPS. You need to set these TLS configurations only if you're using HashiCorp Vault deployments other than HCP Vault Dedicated, like a self-hosted HashiCorp Vault. |
```

josh-wong reviewed (changes requested):
I've added one suggestion. PTAL!

kota2and3kan commented:
@josh-wong  
Thank you for your review!
I applied your suggestion.
Please take a look when you have time! 🙇‍♂️

josh-wong reviewed (approved):
LGTM! Thank you!🙇‍♂️

feeblefakie commented on charts/scalardb-cluster/values.yaml:
```diff
@@ -317,3 +317,17 @@ scalardbCluster:
+    type: ""
+    # -- Vault-specific configurations.
+    vault:
+      # -- TLS configurations to provide access from ScalarDB Cluster to Vault by using TLS. If you're using HashiCorp Cloud Platform (HCP) Vault Dedicated, you don't need to set these TLS configurations even though HCP Vault Dedicated uses HTTPS. You need to set these TLS configurations only if you're using HashiCorp Vault deployments other than HCP Vault Dedicated, like a self-hosted HashiCorp  ...
```
```suggestion
      # -- TLS configurations to provide access from ScalarDB Cluster to Vault by using TLS. If you're using HashiCorp Cloud Platform (HCP) Vault Dedicated, you don't need to set these TLS configurations because HCP Vault Dedicated uses HTTPS. You need to set these TLS configurations only if you're using HashiCorp Vault deployments other than HCP Vault Dedicated, like a self-hosted HashiCorp Vault.
```
`even though` part sounds pretty weird to me.

kota2and3kan replied:
@feeblefakie 
Thank you for your review and discussion in the meeting!
Based on the discussion result, I updated this description in 27b81d9d4242d551b500a7dcc31e999fa1114111.
Please take a look again when you have time! 🙇‍♂️

@josh-wong 
As I described above, I updated this description.
Could you please take a look again when you have time? 🙇‍♂️

josh-wong commented on charts/scalardb-cluster/README.md:
```diff
@@ -25,6 +25,12 @@ Current chart version is `2.0.0-SNAPSHOT`
+| scalardbCluster.encryption.enabled | bool | `false` | Enable encryption at rest. You must set this to `true` if you're using the encryption feature in ScalarDB Cluster. |
+| scalardbCluster.encryption.type | string | `""` | Type of encryption. You must set this value to the same value as "scalar.db.cluster.encryption.type" for ScalarDB Cluster. |
+| scalardbCluster.encryption.vault | object | `{"tls":{"caRootCertSecret":"","enabled":false}}` | Vault-specific configurations. |
+| scalardbCluster.encryption.vault.tls | object | `{"caRootCertSecret":"","enabled":false}` | TLS configurations to provide access from ScalarDB Cluster to Vault by using TLS. If you're using HashiCorp Cloud Platform (HCP) Vault Dedicated, you don't need to set these TLS configurations because HCP Vault Dedicated uses trusted (well-known) CA and ScalarDB Cluster can validate the certificate that  ...
```
```suggestion
| scalardbCluster.encryption.vault.tls | object | `{"caRootCertSecret":"","enabled":false}` | TLS configurations to provide access from ScalarDB Cluster to Vault by using TLS. If you're using HashiCorp Cloud Platform (HCP) Vault Dedicated, you don't need to set these TLS configurations because HCP Vault Dedicated uses a trusted, well-known CA and ScalarDB Cluster can validate the certificate that is provided by HCP Vault Dedicated. You need to set these TLS configurations only if you need to set a custom CA root certificate, for example, you're using your private CA together with HashiCorp Vault deployments other than HCP Vault Dedicated, like a self-hosted HashiCorp Vault. |
```

josh-wong reviewed (approved):
I left just a minor suggestion. Other than that, LGTM! Thank you!🙇‍♂️

feeblefakie reviewed (approved):
LGTM! Thank you!
